### ISSUE TITLE
I don't know how to fix this warning. Getting lots of them

### ISSUE THREAD

**Issue by santanor** [KVault/NESulator#30]
Getting lots of this warnings. Here's the example for one of them

```C:\Users\jossanta\source\repos\NESulator\NESulator\src\gui\../mappers/NROM.h:6:21: warning: its scope is only this definition or declaration, which is probably not what you want [enabled by default]
In file included from C:\Users\jossanta\source\repos\NESulator\NESulator\src\gui\../rom.h:8:0,
                 from C:\Users\jossanta\source\repos\NESulator\NESulator\src\gui\../nes.h:8,
                 from C:\Users\jossanta\source\repos\NESulator\NESulator\src\gui\../memory.h:4,
                 from C:\Users\jossanta\source\repos\NESulator\NESulator\src\gui\nametableViewer.h:5,
                 from C:\Users\jossanta\source\repos\NESulator\NESulator\src\gui\nametableViewer.c:1:
C:\Users\jossanta\source\repos\NESulator\NESulator\src\gui\../mappers/CNROM.h:7:21: warning: 'struct ROM' declared inside parameter list [enabled by default]
 void mapper3(struct ROM *rom);
                     ^

**Comment by the-c0d3r** [comment]
Hi, I stumbled upon this through the "help wanted" sign. This warning message should be solved by adding `struct ROM;` to one line above the `mapper3` function declaration in the header file.

**Comment by santanor** [comment]
@the-c0d3r It worked! (Of course, you knew this already 🥇 )

Thank you so much, I've been googling this for a while but I couldn't get my head around it. Now that being said I still don't get why I need to do that. Is it because of the order in which I've included the headers? 

In any case, you have my kudos, now I don't get an explosion of red every time I compile :D

**Comment by the-c0d3r** [comment]
@santanor No problem. Glad I can help. (sidenote: I'm also quite a beginner in C)

About that warning, based on the message given by gcc, it is taking `struct ROM *rom` as declaring a new struct called ROM with empty body because it haven't seen the forward declaration anywhere yet. It is due to gcc not knowing there is going to be a struct called ROM (aka forward declaration). I guess your header files are not included in correct sequence or variable scoping problem. 

I looked at your header files briefly. I saw that your `CNROM.h` will include `../rom.h` which then include `CNROM.h` again. This might be the issue causing it. You shouldn't include files in this circular order. I usually create a common header file, if the declarations are going to be shared. Or I just add another forward declaration in my header file which uses another header file. And only link them during compiling. 

I think this [SO answer](https://stackoverflow.com/a/16831951/1509809) may be related.
